### PR TITLE
Python2 ==> Python3

### PR DESCRIPTION
**Pull request by Lunarequest**
python 2 has been depreciated on January 1st 2020. I ported the app to python 3. I had to change some sections to use different commands like instead of `.encode('hex')` I had to use binascii.

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -8,157 +10,216 @@
 app.config.from_object("config")
 
 TEMPLATES = {
-    'U':"templateU.bin",
```
Comment (marcan) [inline app.py:11]: What's with all the formatting changes? That doesn't have anything to do with Python 3, right? It makes it hard to see what actually changed.

File: app.py [hunk 2]
```diff
@@ -8,157 +10,216 @@
...
-        oform = {
-            #"privatekey": app.config['RECAPTCHA_PRIVATEKEY'],
-            "secret": app.config['RECAPTCHA_PRIVATEKEY'],
-            "remoteip": request.remote_addr,
-            #"challenge": request.form.get('recaptcha_challenge_field',['']),
-            #"response": request.form.get('recaptcha_response_field',[''])
-            "response": request.form.get('g-recaptcha-response',[''])
+        payload = {
+            "response": request.form.get("g-recaptcha-response", [""]),
+            "secret": app.config["RECAPTCHA_PRIVATEKEY"],
```
Comment (marcan) [inline app.py:103]: Is 'remoteip' not useful for the recaptcha service?

File: app.py [hunk 3]
```diff
@@ -8,157 +10,216 @@
...
 
-    if not any([mac.startswith(i) for i in OUI_LIST]):
-        app.logger.info('Bad MAC %s at %d ver %s bundle %r', mac.encode('hex'), timestamp, request.form['region'], bundle)
-        return _index("The exploit will only work if you enter your Wii's MAC address.")
-
-
-    key = hashlib.sha1(mac+"\x75\x79\x79").digest()
-
-    blob = bytearray(open(os.path.join(app.root_path, template),'rb').read())
+    if not any:
```
Comment (marcan) [inline app.py]: Something went terribly wrong here, this block doesn't make any sense. "not any" is always false.

File: app.py [hunk 4]
```diff
@@ -8,157 +10,216 @@
...
     dt = datetime.utcnow() - timedelta(1)
-    delta = (dt - datetime(2000, 1, 1))
+    delta = dt - datetime(2000, 1, 1)
     timestamp = delta.days * 86400 + delta.seconds
 
     try:
-        mac = ''.join([chr(int(request.form[i],16)) for i in "abcdef"])
-        template = TEMPLATES[request.form['region']]
-        bundle = 'bundle' in request.form
+        mac = "".join([chr(int(request.form[i], 16)) for i in "abcdef"])
```
Comment (marcan) [inline app.py]: This is wrong, mac needs to be `bytes`, not `str`. You are encoding the MAC address as Unicode characters here (chr means different things in py2 and py3).

File: app.py [hunk 5]
```diff
@@ -8,157 +10,216 @@
...
+        template = TEMPLATES[request.form["region"]]
+        bundle = "bundle" in request.form
     except:
         return _index("Invalid input.")
     if not captcha_check():
         return _index("Are you a human?")
 
     if mac == "\x00\x17\xab\x99\x99\x99":
```
Comment (marcan) [inline app.py]: And then this needs to be a bytes comparison.

File: app.py [hunk 6]
```diff
@@ -8,157 +10,216 @@
...
 
-@app.route('/haxx', methods=["POST"])
+
+@app.route("/haxx", methods=["POST"])
 def haxx():
-    OUI_LIST = [i.decode('hex') for i in open(os.path.join(app.root_path, 'oui_list.txt')).read().split("\n") if len(i)==6]
-    g.recaptcha_args = 'k=%s' % app.config['RECAPTCHA_PUBLICKEY']
+
+    OUI_LIST = [
+        binascii.unhexlify(i)
```
Comment (marcan) [inline app.py]: You want `bytes.fromhex()` here, you don't need binascii for anything.

File: app.py [hunk 7]
```diff
@@ -8,157 +10,216 @@
...
+                    "Bad MAC %s at %d ver %s bundle %r",
+                    mac.encode("hex"),
+                    timestamp,
+                    request.form["region"],
+                    bundle,
+                )
+                return _index(
+                    "The exploit will only work if you enter your Wii's MAC address."
+                )
+    mac = bytes(mac, "utf-8")
```
Comment (marcan) [inline app.py]: MAC addresses are definitely not utf-8 encoded characters.

File: app.py [hunk 8]
```diff
@@ -8,157 +10,216 @@
...
-    blob[0x7c:0x80] = struct.pack(">I", timestamp)
-    blob[0x80:0x8a] = "%010d"%timestamp
-    blob[0xb0:0xc4] = hmac.new(key[8:], str(blob), hashlib.sha1).digest()
-
-    path = "private/wii/title/HAEA/%s/%s/%04d/%02d/%02d/%02d/%02d/HABA_#1/txt/%08X.000" % (
-        key[:4].encode('hex').upper(), key[4:8].encode('hex').upper(),
-        dt.year, dt.month-1, dt.day, dt.hour, dt.minute, timestamp
+    blob[0xB0:0xC4] = b"\x00" * 20
+    blob[0x7C:0x80] = struct.pack(">I", timestamp)
+    blob[0x80:0x8A] = bytes("%010d" % timestamp, "utf-8")
```
Comment (marcan) [inline app.py]: `b"%010d" % foo` works, not need to start with a string and then convert.

File: app.py [hunk 9]
```diff
@@ -8,157 +10,216 @@
...
+    blob[0xB0:0xC4] = hmac.new(key[8:], blob, hashlib.sha1).digest()
+
+    path = (
+        "private/wii/title/HAEA/%s/%s/%04d/%02d/%02d/%02d/%02d/HABA_#1/txt/%08X.000"
+        % (
+            binascii.hexlify(key[:4]).upper(),
```
Comment (marcan) [inline app.py]: Just use the `.hex()` method.

File: config.py.sample [hunk 10]
```diff
@@ -9,5 +9,4 @@ RECAPTCHA_PRIVATEKEY = "fillme"
 DEBUG = False
 ADMIN_EMAIL = ['foo@bar.com']
 SMTP_SERVER = '127.0.0.1'
-APP_EMAIL = 'foo@bar.com'
-
+APP_EMAIL = 'foo@bar.com'
```
Comment (marcan) [inline config.py.sample]: Why the lost newline?

**Review by marcan — COMMENTED** [review]
Thanks for the PR! I made a few comments, can you please fix them?

In particular, if you want to reformat the code that's fine, but please squash the change into two commits in that case: one that just reformats and leaves it as python2, and one that actually makes the python3 changes.

File: app.py (hunk 2 above)
Comment (Lunarequest) [reply app.py:103]: its not used. as far as I've seen i get errors when i send it.

File: app.py (hunk 1 above)
Comment (Lunarequest) [reply app.py:11]: oh i think black formater ran on the files. I am very sorry I didn't realize it was active.

File: app.py (hunk 2 above)
Comment (marcan) [reply app.py:103]: It's valid, see [here](https://developers.google.com/recaptcha/docs/verify), and it's working fine in the python2 version. If it's not working for you then there is some other issue.

File: letterbomb.fcgi [hunk 11]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/python2
+#!/usr/bin/python3
```
Comment (marcan) [inline letterbomb.fcgi]: flup does not support Python3, so this whole file needs to be removed, it is useless. I would run this under gunicorn or uwsgi or something like that.

File: app.py (hunk 4 above)
Comment (marcan) [reply app.py]: I think just `bytes(int(request.form[i], 16) for i in "abcdef")` should do the right thing.

File: letterbomb.fcgi (hunk 11 above)
Comment (Lunarequest) [reply letterbomb.fcgi]: I'll add a wsgi file.

File: config.py.sample (hunk 10 above)
Comment (Lunarequest) [reply config.py.sample]: I am not sure I have added it.

File: app.py (hunk 4 above)
Comment (Lunarequest) [reply app.py]: `bytes(int(request.form[i], 16) for i in "abcdef")` does work.

File: app.py (hunk 2 above)
Comment (Lunarequest) [reply app.py:103]: I have added it to the parameters sent.

**Comment by Lunarequest** [comment]
@marcan I have made the changes you requested. everything seems to work

**Comment by Lunarequest** [comment]
@marcan would you mind looking into if this can be merged now.

**Comment by marcan** [comment]
Is this still good? Sorry, I've been pretty busy lately, I didn't intend to ignore this.

**Comment by Lunarequest** [comment]
@marcan please check my new pr

**Pull request closed without merging**